# Homework 4

This is the first homework assignment for which we will be using SageMath.  You first task is to install SageMath on your computer if you haven't done so already.  Then, you have to figure out how to use SageMath as a kernel for your Jupyter Notebook.  Go ahead and give it a shot.  You might have to do a bit of research online to get to this point.  Once you select SageMath as a kernel, you should see SageMath in the top right corner of your notebook instead of just Python.

Several functions that you coded so far in Python have already been coded in SageMath.  For instance, your function `is_prime()` from Homework 1 already exists in SageMath.  So do the functions `ext_euclidean()`, `prime_range()`, `is_prim_root()` from Homework 2, and the function `fast_exp()` from Homework 3.  I encourage you to go ahead and play with big numbers to see how your own code compare to the SageMath code.

In [32]:
p = 1639240535066257
is_prime(p)

True

In [33]:
a = 1134809457890547891543789145789451789145051490847157120934875092387450987230948750923784059872309847609826757134250981723409871238970412307891234078941812
b = 4372894789035250293847509238745097239407592347509872347519087147098134095782093487509872340958723745098723459878901347857213409587234560240567234589723450
xgcd(a,b)

(2,
 -279853842260081926978490496331080850009840587946078747873301328042192450510132403237079577419814578638176116202425600589263595308664451648642204713037479,
 72624840602171283230309502117034901887613599734626470291985001981712468917915291186320198080335808734018015305496554072884997447197162879831497666992131)

In [34]:
L = prime_range(1000000)
len(L)

78498

The prime counting function 
$$\pi(x) = \# \{p \text{ prime } : p \le x \}$$
that we talked about in class is also implemented in SageMath.  Thus, the number above could have been calculated as follows instead:

In [35]:
prime_pi(1000000)

78498

In SageMath, modular arithmetic can be dealt with as follows:

In [36]:
print(Mod(3,15) + Mod(23,15))
print(Mod(3,15) * Mod(23,15))
print(Mod(4,15)^(-1))

11
9
4


And one can ask Sage if a given integer modulo m is a primitive root:

In [37]:
Mod(2,5).is_primitive_root()

True

Or asks for the multiplicative order of a unit modulo m:

In [38]:
Mod(2,5).multiplicative_order()

4

Of course, the Euler phi function is implemented in SageMath:

In [39]:
euler_phi(100)

40

Fast modular exponentiation is already implemented in Python when working with integers modulo m:

In [40]:
Mod(2,5)^10394857190348751

3

Another function you haven't implemented yourself that can be handy to know is the function `factor()`:

In [41]:
factor(109238741029348749)

3^3 * 7 * 13 * 41 * 2237 * 484754321

## Problem 1

Using SageMath now, compute the ten smallest primitive roots modulo the prime:

In [42]:
p = 6242701953154426055792104938207719351221

In [43]:
prim_roots = []
m = 2

while len(prim_roots) < 10:
    if Mod(m, p).is_primitive_root():
        prim_roots.append(m)
    m += 1

prim_roots

[2, 6, 11, 19, 24, 26, 30, 33, 34, 37]

## Problem 2

SageMath uses different types than Python.  For instance:

In [44]:
a = 2
print(type(2))

b = 2/3
print(type(2/3))

<class 'sage.rings.integer.Integer'>
<class 'sage.rings.rational.Rational'>


But one can cast those types into the familiar Python ones if desired:

In [45]:
type(int(a))

<class 'int'>

For the following problem, you will need to cast numbers as real numbers which can be done as follows:

In [46]:
print(RR((a)))
print(RR((b)))

2.00000000000000
0.666666666666667


It is conjectured that $2$ is a primitive root for infinitely many primes $p$, but no one can prove this...This is a special case of [Artin's primitive root conjecture](https://en.wikipedia.org/wiki/Artin%27s_conjecture_on_primitive_roots).  The following real number is known as Artin's constant:
$$A = \lim_{x \to \infty} \frac{\# \{p \le x : 2 \text{ is a primitive root modulo } p \}}{\# \{p \le x \}}. $$
Estimate the value of Artin's constant numerically by computing the quotient appearing above for $x = 10^{k}$ for $k=1,2,3,4,5,6,7$ and casting it to a real number.  

In [47]:
for k in range(1, 8):
    x = 10^k
    
    temp_primes = prime_range(2, x + 1)
    total_primes = len(temp_primes)
    
    prim_count = 0
    for p in temp_primes:
        if Mod(2, p).is_primitive_root():
            prim_count += 1
    
    ratio = RR(prim_count) / RR(total_primes)
    print("k =", k, "x =", x, "ratio =", ratio)

k = 1 x = 10 ratio = 0.500000000000000
k = 2 x = 100 ratio = 0.480000000000000
k = 3 x = 1000 ratio = 0.398809523809524
k = 4 x = 10000 ratio = 0.382424735557364
k = 5 x = 100000 ratio = 0.375625521267723
k = 6 x = 1000000 ratio = 0.373780223699967
k = 7 x = 10000000 ratio = 0.373907390994901


## Problem 3

In class, we defined a safe prime to be a prime $p$ of the form $p = 2q + 1$ for some other primes $q$.  (The primes $q$ for which $2q + 1$ is prime are called  Sophie Germain primes.)  Calculate how many safe primes there are below $10^4$.

In [48]:
total = 0

for p in prime_range(2, 10^4):
    q = (p - 1) // 2
    if is_prime(q):
        total += 1

total

115

## Problem 4

Implement the naive way to solve the discrete log problem
$$g^{x} \equiv h \pmod{p} $$
as we explained in class.  Here $g$ is any unit modulo $p$, $h$ any integer and we take $p$ to be a prime.  The function call should look like
```
def naive_dl(g,p,h):
    list_of_powers = list()
    [COMPLETE]
    return x
```

In [49]:
def naive_dl(g,p,h):
    list_of_powers = list()

    h = Mod(h, p)
    
    for x in range(p):
        list_of_powers.append(Mod(g, p)^x)
    
    for x in range(len(list_of_powers)):
        if list_of_powers[x] == h:
            return x

## Problem 5

Now, implement Shanks's Babystep-Giantstep Algorithm for computing the discrete log.  The function call should look like:
```
def sbsgs_dl(g,p,h):
    n = 1 + floor(sqrt(p-1))
    [COMPLETE]
    y = i + j*n
    return y
```

In [50]:
def sbsgs_dl(g, p, h):
    n = 1 + floor(sqrt(p - 1))
    g = Mod(g, p)
    h = Mod(h, p)
    
    baby_steps = []
    for i in range(n):
        baby_steps.append(g^i)

    giant_steps = []
    for j in range(n):
        giant_steps.append(h * g^(-j*n))

    for i in range(n):
        for j in range(n):
            if baby_steps[i] == giant_steps[j]:
                y = i + j*n
                return y
            

## Problem 6

Now compare both ways as follows.  We will test your functions by timing them on a $28$ bit prime.  (We already talked about Fermat pseudoprimes, and we'll talk more about pseudoprimes later in this class.)  Once your functions `naive_dl()` and `sbsgs_dl()` work fine, run the following code:

In [51]:
# Return a pseudoprime with k bits
def get_pseudoprime(k):
    p = 1
    while not is_pseudoprime(p):
        p = ZZ(randint(2^(k-1),2^k))
    return p

# Time your function with a 28 bit prime
k = 28
p = get_pseudoprime(k)
g = primitive_root(p)
h = 123

Make sure the cell just above ran, and now run the following code in the cell below:
```
time x = naive_dl(g,p,h)
```

In [52]:
time x = naive_dl(g,p,h)

CPU times: user 3min 29s, sys: 42 s, total: 4min 11s
Wall time: 4min 32s


As a check, run also the following code in the cell below:
```
print(g,p,x)
print(Mod(g,p)^x == h)
```

In [53]:
print(g,p,x)
print(Mod(g,p)^x == h)

3 206616847 46751748
True


Run the following code in the cell below:
```
time y = sbsgs_dl(g,p,h)
```

In [54]:
time y = sbsgs_dl(g,p,h)

CPU times: user 1.79 s, sys: 5.59 ms, total: 1.79 s
Wall time: 1.8 s


As a check, run also the following code in the cell below:
```
print(g,p,y)
print(Mod(g,p)^y == h)
```

In [55]:
print(g,p,y)
print(Mod(g,p)^y == h)

3 206616847 46751748
True


How do the running times compare between the naive approach and the Babystep-Giantstep approach?

The naive approach took about 4 minutes while the Babystep-Giantstep approach took only 1.8 seconds. That's a huge difference in runtime when it begins to scale.

In [56]:
k = 32
p = get_pseudoprime(k)
g = primitive_root(p)
h = 123

In [57]:
time y = sbsgs_dl(g,p,h)

CPU times: user 3.74 s, sys: 9.03 ms, total: 3.75 s
Wall time: 3.77 s


In [58]:
print(g,p,y)
print(Mod(g,p)^y == h)

5 3265426847 3095378456
True


In [59]:
k = 64
p = get_pseudoprime(k)
g = primitive_root(p)
h = 123

In [ ]:
time y = sbsgs_dl(g,p,h)

In [ ]:
print(g,p,y)
print(Mod(g,p)^y == h)

Try out the Babystep-Giantstep approach with a $32$ bit prime, then a $64$ bit prime.  What is the largest $k$ you can "crack"?

I was able to crack the 32 bit prime in 3.75 seconds but as of 23 minutes the 64 bit prime is still going and I decided to stop it